File: 1715-split-a-string-into-the-max-number-of-unique-substrings/1715-split-a-string-into-the-max-number-of-unique-substrings.cs
public class Solution {
    public int MaxUniqueSplit(string s) {
        var rs = MaxUniqueSplit(0, new HashSet<string>(), s);
        return rs;
    }
    private int MaxUniqueSplit(int index, HashSet<string> set, string s) {
        var rs = 0;
        if (index >= s.Length) return rs;
        for (int i = index; i < s.Length; i++) {
            var part = s.Substring(index, i - index + 1);
            if (!set.Contains(part)) {
                set.Add(part);
                var rs0 = 1 + MaxUniqueSplit(i + 1, set, s);
                rs = Math.Max(rs, rs0);
                set.Remove(part);
            }
        }
        return rs;
    }
}

File: 1715-split-a-string-into-the-max-number-of-unique-substrings/1715-split-a-string-into-the-max-number-of-unique-substrings.java
class Solution {
    private char[] cs;
    private int n;
    private int res;

    public int maxUniqueSplit(String s) {
        cs = s.toCharArray();
        n = cs.length;
        dfs(0, 0, new HashSet<>());
        return res;
    }

    private void dfs(int hash, int i, Set<Integer> vis) {
        if (i == n) {
            res = Math.max(res, vis.size());
            return;
        }
        hash = hash * 131 + cs[i];
        if (!vis.contains(hash)) {
            vis.add(hash);
            dfs(0, i + 1, vis);
            vis.remove(hash);
        }
        dfs(hash, i + 1, vis);
    }

}